modules/darwinosl/darwinosl-source.m

```diff
@@ -0,0 +1,575 @@
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+
+  if (self->stat_persist_name == NULL)
+    self->stat_persist_name = _generate_persist_name(self, "darwinosl", s->super.super.super.persist_name);
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config((LogPipe *)self);
+
+  darwinosl_source_persist_init(self->log_source_persist, cfg->state, _get_persist_name(s));
+
+  gboolean ret = log_threaded_source_driver_init_method(s);
+  if (ret)
+    _register_aggregated_stats(self);
+
+  return ret;
+}
+
+static gboolean
+_deinit(LogPipe *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+
+  _unregister_aggregated_stats(self);
+
+  return log_threaded_source_driver_deinit_method(s);
+}
+
+static void
+_free(LogPipe *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+
+  /* NOTE: ARC takes care of the rest */
+  self->osLogSource = nil;
+
+  darwinosl_sd_options_destroy(&self->options);
+  darwinosl_source_persist_free(self->log_source_persist);
+
+  if (self->persist_name)
+    g_free((gpointer)self->persist_name);
+  if (self->stat_persist_name)
+    g_free((gpointer)self->stat_persist_name);
+
+  log_threaded_source_driver_free_method(s);
+}
+
+static LogThreadedSourceWorker *
+_construct_worker(LogThreadedSourceDriver *s, gint worker_index)
+{
+  /* TODO: DarwinOSLogSourceDriver uses the multi-worker API, but it is not prepared now to work with more than one worker. */
+  g_assert(s->num_workers == 1);
+
+  LogThreadedSourceWorker *worker = g_new0(LogThreadedSourceWorker, 1);
+  log_threaded_source_worker_init_instance(worker, s, worker_index);
+
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *) s;
+  g_atomic_counter_set(&self->exit_requested, FALSE);
+  worker->run = _run;
+  worker->request_exit = _request_exit;
+
+  return worker;
+}
+
+LogDriver *
+darwinosl_sd_new(GlobalConfig *cfg)
+{
+  DarwinOSLogSourceDriver *self = g_new0(DarwinOSLogSourceDriver, 1);
+  log_threaded_source_driver_init_instance(&self->super, cfg);
+
+  self->super.super.super.super.init = _init;
+  self->super.super.super.super.deinit = _deinit;
+  self->super.super.super.super.free_fn = _free;
+  self->super.super.super.super.generate_persist_name = _get_persist_name;
+
+  self->super.worker_options.ack_tracker_factory = consecutive_ack_tracker_factory_new();
+  self->super.worker_construct = _construct_worker;
+
+  self->super.format_stats_key = _format_stats_key;
+
+  self->log_source_persist = darwinosl_source_persist_new();
+
+  self->osLogSource = [OSLogSource new];
+
+  darwinosl_sd_options_defaults(&self->options, &self->super.worker_options);
+
+  return &self->super.super.super;
+}
+
+/* Options */
+
+gboolean
+darwinosl_sd_set_filter_predicate(LogDriver *s, const gchar *filter_predicate_str)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
```

Comment:
Check the predicate syntax by try to initialize it or something.

Currently if the syntax is invalid for example `filter-predicate("adassdadad")` than the exception will not be catched and the ObjC exception will kill the process, but if the predicate is only invalid in terms of semantics like `filter-predicate("adassdadad == true")` syslog-ng wil warn you about it like `[2024-01-10T16:36:00.574676] darwinosl: Could not get OSLog enumerator; error_code='-1', error='issue with predicate: no such field: adassdadad'` but it will continue executing.

By checking the predicate's format here will catch this issue, and maybe we can even show where the error is inside the configuration file for better user experience when having >1 darwing-oslog sources

Reply:
Seems there is no chance to catch the NSInvalidArgumentException, so this will still always crash if the predicte string has certain parsing issues. I concider this an Apple bug, trying to get a workaround of this later.